docs/source/development/eep-02-typing.md

```diff
@@ -0,0 +1,1479 @@
+
+In the long run we plan a general overhaul of `MSM` estimation that provides better
+access to currently internal objects such as the MSM objective function.
+
+(design-philosophy)=
+
+## Design Philosophy
+
+- User facing functions should be generous regarding their input type. Example: the
+  `algorithm` argument can be a string, `Algorithm` class or `Algorithm` instance. The
+  `algo_options` can be an `AlgorithmOptions` object or a dictionary of keyword
+  arguments.
+- User facing functions should be strict about their output types. A strict output type
+  does not just mean that the output type is known (and not a generous Union), but that
+  it is a proper type that enables static analysis for available attributes. Example:
+  whenever possible, public functions should not return dicts but proper result types
+  (e.g. `OptimizeResult`, `NumdiffResult`, ...)
+- Internal functions should be strict about input and output types; Typically, a public
+  function will check all arguments, convert them to a proper type and then call an
+  internal function. Example: `minimize` will convert any valid value for `algorithm`
+  into an `Algorithm` instance and then call an internal function with that type.
+- Each argument that previously accepted strings or option dictionaries now also accepts
+  input types that are more amenable to static analysis and offer better autocomplete.
+  Example: `algo_options` could just be a dict of keyword arguments. Now it can also be
+  an `AlgorithmOptions` instance that enables autocomplete and static analysis for
+  attribute access.
+- Fixed field types should only be used if all fields are known. An example where this
+  is not the case are benchmark results, that are better represented as dictionaries.
+- The documentation will mainly show preferred ways of doing things. Example: while
+  option dictionaries are allowed in many cases, we always show how to do things in more
+  autocomplete friendly ways.
+- Whenever possible, use immutable types. Whenever things need to be changeable,
+  consider using an immutable type with copy constructors for modified instances.
+  Example: instances of `Algorithm` are immutable but using `Algorithm.with_option`
+  users can create modified copies.
+- The main entry point to estimagic are functions, objects are mostly used for
+  configuration and return types. This takes the best of both worlds: we get the safety
+  and static analysis that (in Python) can only be achieved using objects but the
+  beginner friendliness and freedom provided by functions. Example: Having a `minimize`
+  function, it is very easy to add the possibility of running minimizations with
+  multiple algorithms in parallel and returning the best value. Having a `.solve` method
+  on an algorithm object would require a whole new interface for this.
+
+## Type checkers and their configuration
+
+We choose mypy as static type checker and run it as part of our continuous integration.
```

Comment:
We could also think about running mypy using pre-commit hooks if it's not too slow. This is what JAX is doing (see their [.pre-commit-config.yaml](https://github.com/google/jax/blob/main/.pre-commit-config.yaml).

Reply:
Added it.